### PR TITLE
Fix traffic split issue

### PR DESCRIPTION
**Pull request by StrongMonkey**

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rio/issues/485

File: modules/gateway/controllers/routeset/populate/virtualservice.go [hunk 1]
```diff
@@ -6,6 +6,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/rancher/mapper/slice"
+
```
Comment (daxmc99) [inline modules/gateway/controllers/routeset/populate/virtualservice.go]: nit: formatting

File: modules/gateway/controllers/routeset/populate/virtualservice.go [hunk 2]
```diff
@@ -66,16 +68,14 @@ func virtualServiceFromRoutesets(systemNamespace string, clusterDomain *projectv
 				port = int(*dest.Port)
 			}
 
-			if i == len(routeSpec.To)-1 && constants.ServiceMeshMode == constants.ServiceMeshModeLinkerd {
-				host := dest.Service
-				if dest.Revision != "" {
-					host = host + "-" + dest.Revision
+			if constants.ServiceMeshMode == constants.ServiceMeshModeLinkerd {
+				httpRoute.Headers.Request.Set["l5d-dst-override"] = fmt.Sprintf("%s-%v.%s.svc.cluster.local:%v", routeSet.Name, index, routeSet.Namespace, port)
```
Comment (daxmc99) [inline modules/gateway/controllers/routeset/populate/virtualservice.go]: I feel like we need to mention something about https://github.com/linkerd/linkerd2-proxy/pull/173
Or document what `l5d-dst-override` does. I'm even fine with a link the linkerd2 PR. I feel like this is hard to understand without some more context with a comment

**Comment by StrongMonkey** [comment]
This pr is to fix two things:
1. TrafficSplit on routeset should work. This is controlled by overriding header in virtualservice with 
routeName-${index}.namespace.svc.cluster.local. (Index is the number of route in router),  for each index a trafficSplit is created. 
2. TrafficSplit on app(aggregation of services) is done by overriding header in virtual service with ${app_name}.namespace.svc.cluster.local. That's why we need aggregated  flag to control whether we write version or no.

File: modules/gateway/controllers/service/populate/virtualservice.go [hunk 3]
```diff
@@ -264,7 +264,7 @@ func isProtocolSupported(protocol v1.Protocol) bool {
 }
 
 func VirtualServiceFromSpec(aggregated bool, systemNamespace string, name, namespace string, clusterDomain *projectv1.ClusterDomain, service *v1.Service, dests ...Dest) *v1alpha3.VirtualService {
```
Comment (daxmc99) [inline modules/gateway/controllers/service/populate/virtualservice.go]: Could we add a comment to define what `aggregated` does?

Comment (daxmc99) [reply modules/gateway/controllers/service/populate/virtualservice.go]: My understanding is that aggregated determines if its the route to a revision versus the "app" itself.

File: modules/gateway/controllers/service/populate/virtualservice.go [hunk 4]
```diff
@@ -100,7 +100,7 @@ func httpRoutes(systemNamespace string, service *v1.Service, dests []Dest) ([]v1
 
 	return result, external
 }
-func newRoute(systemNamespace, externalGW string, published bool, portBinding v1.ContainerPort, dests []Dest, appendHTTPS bool, autoscale bool, svc *v1.Service) (string, v1alpha3.HTTPRoute) {
+func newRoute(aggregated bool, systemNamespace, externalGW string, published bool, portBinding v1.ContainerPort, dests []Dest, appendHTTPS bool, autoscale bool, svc *v1.Service) (string, v1alpha3.HTTPRoute) {
```
Comment (daxmc99) [inline modules/gateway/controllers/service/populate/virtualservice.go]: Maybe the explanation would be better here?

**Comment by StrongMonkey** [comment]
@daxmc99 Please review, comments has been addressed.

**Pull request merged**